### PR TITLE
[#79] Make sure that suggestion doesn't remove any parts of the line

### LINKED ISSUE #79: Make sure that suggestion doesn't remove any parts of the line
In HLint the `Idea` stores the `SrcLoc` of the corresponding code of the hint, so we should make sure that we make the suggestion only to the corresponding part and leave the rest as it is.

### PR DESCRIPTION
**Pull request by vrom911**
Resolves #79

### CODE REVIEW

File: src/Hintman/Core/Review.hs [hunk 1]
```diff
@@ -44,7 +44,7 @@ data Comment = Comment
 
       -- | Text of the review comment
     , commentBody     :: !Text
-    }
+    } deriving (Show, Eq)
```
Comment (chshersh) [inline src/Hintman/Core/Review.hs]: I know what we are going to do when we will support GHC-8.8 :smirk:

File: src/Hintman/HLint.hs [hunk 2]
```diff
@@ -92,28 +108,62 @@ createFileDownloadUrl PrInfo{..} (toText -> file) = T.intercalate "/"
...
     -- This function is copy-pasted from HLint sources.
     showNotes :: [Note] -> Text
     showNotes = T.intercalate ", " . map show . filter use
       where
         use :: Note -> Bool
         use ValidInstance{} = False -- Not important enough to tell an end user
         use _               = True
+
+-- | Safe @at@ function.
+(!!?) :: [a] -> Int -> Maybe a
```
Comment (chshersh) [inline src/Hintman/HLint.hs]: Nice function!

File: src/Hintman/HLint.hs [hunk 3]
```diff
@@ -92,28 +108,62 @@ createFileDownloadUrl PrInfo{..} (toText -> file) = T.intercalate "/"
...
+{- | Creates the comment text from the HLint 'Idea's and file content.
 -}
-createComment :: Idea -> Maybe Text
-createComment Idea{..} = ideaTo >>= \to ->
+createCommentText :: Text -> Idea -> Maybe Text
+createCommentText content Idea{..} = ideaTo >>= \to ->
     if ideaSeverity == Ignore
     then Nothing
-    else Just $ unlines $
+    else buildWholeLine to >>= \newLine -> Just $ unlines $
```
Comment (chshersh) [inline src/Hintman/HLint.hs]: We can rewrite this code in a slightly cleaner way using `do`-notation for `Maybe` if you think it's better:

```haskell
createCommentText content Idea{..} = do
    to <- ideaTo
    guard $ ideaSeverity /= Ignore
    newLine <- buildWholeLine to
    ...
```

File: test/Test/HLint.hs [hunk 4]
```diff
@@ -16,22 +17,11 @@ hlintSpec :: Spec
 hlintSpec = describe "HLint works on opened PRs" $ do
     it "works with non-code PRs" $
         pr1 >>= runLog . runHLint >>= shouldBe []
-    it "works on code PRs" $
-        pr2 >>= runLog . runHLint >>= shouldBe [etaReduce] . map show
     it "creates correct comment for PR 2" $
-        pr2 >>= runLog . runHLint >>= shouldBe [etaComment] . map createComment
+        pr2 >>= runLog . runHLint >>= shouldBe [etaComment] . map commentBody
```
Comment (chshersh) [inline test/Test/HLint.hs]: Let's also add a test for a hint that doesn't change the whole line, only a part :slightly_smiling_face:

File: src/Hintman/HLint.hs (hunk 3 above)
Comment (vrom911) [reply src/Hintman/HLint.hs]: Nice! Much cleaner!

File: src/Hintman/HLint.hs [hunk 5]
```diff
@@ -92,28 +108,62 @@ createFileDownloadUrl PrInfo{..} (toText -> file) = T.intercalate "/"
...
         (if ideaHint == "" then "" else show ideaSeverity <> ": " <> toText ideaHint)
```
Comment (vrom911) [inline src/Hintman/HLint.hs]: @chshersh maybe we should write the `ideaSeverity` anyway without this check?

Comment (chshersh) [reply src/Hintman/HLint.hs]: @vrom911 No, I think that the current behavior is okay. If something is ignored, people probably don't want to see it. But later we can implement `ShowIgnore` flag that will display ignores as well if somebody wants.

File: test/Test/HLint.hs [hunk 6]
```diff
@@ -16,24 +17,51 @@ hlintSpec :: Spec
...
-    it "creates correct comment for PR 2" $
-        pr2 >>= runLog . runHLint >>= shouldBe [etaComment] . map createComment
+    it "creates correct eta-reduce comment for PR 2" $
+        pr2 >>= runLog . runHLint >>= (`shouldSatisfy` (etaComment `elem`))
+    it "creates correct part line comment for PR 2" $
+        pr2 >>= runLog . runHLint >>= (`shouldSatisfy` (avoidLambdaComment `elem`))
+    it "creates remove line comment for PR 2" $
+        pr2 >>= runLog . runHLint >>= (`shouldSatisfy` (removePragmaComment `elem`))
+    it "creates multiline comment for PR 2" $
+        pr2 >>= runLog . runHLint >>= (`shouldSatisfy` (multilineComment `elem`))
```
Comment (chshersh) [inline test/Test/HLint.hs]: Tests are :top:  :100:

**Pull request merged**